CrossCorrelate and GetDetectorOffset algorithm user docs (#38538)
mantidproject/mantid: docs/source/algorithms/CrossCorrelate-v1.rst

Description
-----------

Compute the cross correlation function for a range of spectra with respect to a reference spectrum.

> This is use in powder diffraction experiments when trying to estimate the offset (in number of bin indices) of one spectra with respect to another one.

Reviewer: ```suggestion
This is used in powder diffraction experiments when trying to estimate the offset (in number of bin indices) of one spectrum with respect to another one, where the x-axis units share a common crystallographic unit (such as d-spacing or momentum transfer).
```

> The input spectra should be in d-spacing and then interpolate on the X-axis of the reference before calculating the cross correlation in bin index.

Reviewer: In principle, this should work with momentum transfer too, but I can't think of a practical situation where someone would do this.  Hence the suggested change.

Author: In the algorithm, there is a check for the units on the input workspace to be d-spacing. I agree that it should work on Q as well, but haven't ever edited the code so I could try.

> The input spectra should be in d-spacing and then interpolate on the X-axis of the reference before calculating the cross correlation in bin index.

Reviewer: ```suggestion
The input spectra should (normally) be in d-spacing and then interpolated on the X-axis of the reference before calculating the cross correlation in bin index.
```

> The cross correlation function is computed

Reviewer: ```suggestion
The cross correlation function is computed as:
```

.. math:: c_{fg}(r) = \frac{1}{\sigma_f \sigma_g} \int_{XMin}^{XMax} (f(x) - \mu_f) (g(x+r)-\mu_g) dx

where :math:`x` and :math:`r` are bin indices.

> :math:`r` being either half of the number of points in the requested range, ``XMin`` to ``XMax``, of the reference spectrum, or up to `MaxDSpaceShift`, whichever is smaller.

Reviewer: since `r` is in bins and `MaxDSpaceShift` in d-spacing, I think a couple of words extra are needed here:

> :math:`r` being either half of the number of points in the requested range, ``XMin`` to ``XMax``, of the reference spectrum, or up to `MaxDSpaceShift`, whichever is smaller.

Reviewer: ```suggestion
:math:`r` being either half of the number of points in the requested range, ``XMin`` to ``XMax``, of the reference spectrum, or up to the number of bins corresponding to `MaxDSpaceShift`, whichever is smaller.
```

> :math:`f(x)` is the reference spectrum, with :math:`\mu_f` and :math:`\sigma_f` being the mean and variance respectively.

Reviewer: mean and variance make sense for input Gaussian spectra and also close enough to make sense when it's a single TOF-peak, but what do they mean when the input is an entire powder pattern?

Author: It really is the "simple" mean and variance of the y-values. It is a way to make the cross-correlation value more behaved. Those equations can be added or links for the pages in Wikipedia, but I'm not sure that would help people understand. Do you have different wording that can help?

Reviewer: not really.

Comments for creating useful cross-correlation output
#####################################################

* Supply input data with relatively high signal to noise ratio.
  When this is not met, the resulting cross-correlation generally have issues having offsets determined from them by :ref:`algm-GetDetectorOffsets`.
* Take note of input binning structure before using :ref:`algm-GetDetectorOffsets` and :ref:`algm-ConvertDiffCal`.
  The different binning modes (linear or logorithmic) correspond to different behavior in the follow-on algorithms.
* Cross-correlating linear binning generally uses ``OffsetMode="Relative"`` or ``OffsetMode="Absolute"`` and requires a narrow range of integration around a single Bragg peak with known reference d-spacing.
* Use a large range for ``XMin`` to ``XMax``.
  This will promote calibrations that take a larger range of the data into account rather than a single diffraction peak.
  However, it only works for logorithmic binning which will use ``OffsetMode="Signed"`` in :ref:`algm-GetDetectorOffsets`.
* With wide range of integration, use the ``MaxDSpaceShift`` to limit the value of :math:`r`.

>   Not only will this speed up the calculation, but the output spectra will only include where the spectra are similar to the reference and not have spurious peaks or background.

Reviewer: ```suggestion
  Not only will this speed up the calculation, but the output spectra will only include where the spectra are similar to the reference and not have spurious peaks or background (but note: `MaxDSpaceShift` must be greater than  approximately double the maximum Bragg peak full-width of the input data to ensure that that the cross correlation contains a complete peak).
```

Description
-----------

Compute the cross correlation function for a range of spectra with respect to a reference spectrum.

This is used in powder diffraction experiments when trying to estimate the offset (in number of bin indices) of one spectrum with respect to another one, where the x-axis units share a common crystallographic unit (such as d-spacing or momentum transfer).
The input spectra need to be in d-spacing and then interpolated on the X-axis of the reference before calculating the cross correlation in bin index.
In principle, momentum transfer will work, but this is not currently supported.
The cross correlation function is computed as:

> .. math:: c_{fg}(r) = \frac{1}{\sigma_f \sigma_g} \int_{XMin}^{XMax} (f(x) - \mu_f) (g(x+r)-\mu_g) dx

Author: I loathe bar notation because typesetting can often make it nearly indecipherable when exponents are involved. It has caused a couple of decades of confusion in the total scattering community.

> .. math:: c_{fg}(r) = \frac{1}{\sigma_f \sigma_g} \int_{XMin}^{XMax} (f(x) - \mu_f) (g(x+r)-\mu_g) dx

Reviewer 2: $\bar{f^2}$ vs $\bar{f}^2$.  What's the issue?

  When this is not met, the resulting cross-correlation generally have issues having offsets determined from them by :ref:`algm-GetDetectorOffsets`.
* Take note of input binning structure before using :ref:`algm-GetDetectorOffsets` and :ref:`algm-ConvertDiffCal`.
  The different binning modes (linear or logorithmic) correspond to different behavior in the follow-on algorithms.
* Cross-correlating linear binning generally uses ``OffsetMode="Relative"`` or ``OffsetMode="Absolute"`` and requires a narrow range of integration around a single Bragg peak with known reference d-spacing.
* Use a large range for ``XMin`` to ``XMax``.
  This will promote calibrations that take a larger range of the data into account rather than a single diffraction peak.
  However, it only works for logorithmic binning which will use ``OffsetMode="Signed"`` in :ref:`algm-GetDetectorOffsets`.
* With wide range of integration, use the ``MaxDSpaceShift`` to limit the value of :math:`r`.
  Not only will this speed up the calculation, but the output spectra will only include where the spectra are similar to the reference and not have spurious peaks or background.

>   **NOTE:** `MaxDSpaceShift` must be greater than  approximately double the maximum Bragg peak full-width of the input data to ensure that that the cross correlation contains a complete peak.

Reviewer: ```suggestion
  **NOTE:** `MaxDSpaceShift` must be greater than  approximately double the maximum Bragg peak full-width of the input data to ensure that the cross correlation contains a complete peak.
```
